source/Wormhole.cpp

```diff
@@ -0,0 +1,207 @@
+		bool hasValue = (child.Size() > valueIndex);
+		const string &value = child.Token(hasValue ? valueIndex : 0);
+		
+		// Check for conditions that require clearing this key's current value.
+		// "remove <key>" means to clear the key's previous contents.
+		// "remove <key> <value>" means to remove just that value from the key.
+		if((remove && !hasValue) && key == "link")
+		{
+			links.clear();
+			continue;
+		}
+		
+		// Handle the attributes which can be "removed."
+		if(key == "link" && child.Size() > valueIndex + 1)
+		{
+			const auto *from = GameData::Systems().Get(child.Token(valueIndex));
+			const auto *to = GameData::Systems().Get(child.Token(valueIndex + 1));
+			if(remove)
+			{
+				// Only erase if the link is an exact match.
+				auto it = links.find(from);
+				if(it != links.end() && it->second == to)
+					links.erase(from);
+			}
+			else
+				links[from] = to;
+		}
+		else if(key == "linked")
+			linked = !remove;
+		else if(key == "display name")
+		{
+			if(remove)
+				name = "???";
+			else if(hasValue)
+				name = value;
+		}
+		else if(remove)
+			child.PrintTrace("Cannot \"remove\" a specific value from the given key:");
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+	}
+}
+
+
+
+void Wormhole::LoadFromPlanet(const Planet *planet)
+{
+	this->planet = planet;
+	linked = !planet->Description().empty();
+	GenerateLinks();
+	isAutogenerated = true;
+	isDefined = true;
+}
+
+
+
+bool Wormhole::IsValid() const
+{
+	if(!isDefined)
+		return false;
+
+	for(auto &&pair : links)
+		if(!pair.first->IsValid() || !pair.second->IsValid())
+			return false;
+
+	return true;
+}
+
+
+
+const Planet *Wormhole::GetPlanet() const
+{
+	return planet;
+}
+
+
+
+const string &Wormhole::Name() const
+{
+	return name;
+}
+
+
+
+bool Wormhole::IsLinked() const
+{
+	return linked;
+}
+
+
+
+bool Wormhole::IsAutogenerated() const
+{
+	return isAutogenerated;
+}
+
+
+
+const System *Wormhole::WormholeSource(const System *to) const
+{
+	using value_type = decltype(links)::value_type;
+	auto it = find_if(links.begin(), links.end(),
+			[&to](const value_type &val)
+			{
+				return val.second == to;
+			});
+
+	auto source = it == links.end() ? to : it->first;
+
+	// If source refers to a system without this wormhole, go through the link backwards.
+	if(it != links.end() && !planet->IsInSystem(source))
+		source = WormholeSource(source);
+	return source;
+}
+
+
+
+
+const System *Wormhole::WormholeDestination(const System *from) const
+{
+	auto it = links.find(from);
+	auto dest = it == links.end() ? from : it->second;
+
+	// If dest refers to a system without this wormhole, go through the link.
+	if(it != links.end() && !planet->IsInSystem(dest))
+		dest = WormholeDestination(dest);
+	return dest;
+}
+
+
+
+const unordered_map<const System *, const System *> &Wormhole::Links() const
+{
+	return links;
+}
+
+
+
+void Wormhole::SetLinked(bool linked)
+{
+	this->linked = linked;
+}
+
+
+
+void Wormhole::GenerateLinks()
```

Comment:
Would it make sense to make the code in this function part of `LoadFromPlanet()`? (Or at least rename it to `"GenerateLinksFromPlanets()`).